message/message.go

```diff
@@ -0,0 +1,285 @@
+package message
+
+import (
+	"fmt"
+
+	ggio "github.com/gogo/protobuf/io"
+	pb "github.com/ipfs/go-graphsync/message/pb"
+	gsselector "github.com/ipfs/go-graphsync/selector"
+	ipld "github.com/ipfs/go-ipld-format"
+)
+
+// GraphSyncRequestID is a unique identifier for a GraphSync request.
+type GraphSyncRequestID int32
+
+// GraphSyncPriority a priority for a GraphSync request.
+type GraphSyncPriority int32
+
+// GraphSyncResponseStatusCode is a status returned for a GraphSync Request.
+type GraphSyncResponseStatusCode int32
+
+const (
+
+	// GraphSync Response Status Codes
+
+	// Informational Response Codes (partial)
+
+	// RequestAcknowledged means the request was received and is being worked on.
+	RequestAcknowledged = 10
+	// AdditionalPeers means additional peers were found that may be able
+	// to satisfy the request and contained in the extra block of the response.
+	AdditionalPeers = 11
+	// NotEnoughGas means fulfilling this request requires payment.
+	NotEnoughGas = 12
+	// OtherProtocol means a different type of response than GraphSync is
+	// contained in extra.
+	OtherProtocol = 13
+
+	// Success Response Codes (request terminated)
+
+	// RequestCompletedFull means the entire fulfillment of the GraphSync request
+	// was sent back.
+	RequestCompletedFull = 20
+	// RequestCompletedPartial means the response is completed, and part of the
+	// GraphSync request was sent back, but not the complete request.
+	RequestCompletedPartial = 21
+
+	// Error Response Codes (request terminated)
+
+	// RequestRejected means the node did not accept the incoming request.
+	RequestRejected = 30
+	// RequestFailedBusy means the node is too busy, try again later. Backoff may
+	// be contained in extra.
+	RequestFailedBusy = 31
+	// RequestFailedUnknown means the request failed for an unspecified reason. May
+	// contain data about why in extra.
+	RequestFailedUnknown = 32
+	// RequestFailedLegal means the request failed for legal reasons.
+	RequestFailedLegal = 33
+	// RequestFailedContentNotFound means the respondent does not have the content.
+	RequestFailedContentNotFound = 34
+)
+
+// GraphSyncRequest is an interface for accessing data on request contained in a
+// GraphSyncMessage.
+type GraphSyncRequest interface {
+	Selector() gsselector.Selector
+	Root() ipld.Node
```

Comment:
why is root a node and not a cid?

Reply:
Doing it this way means we have to first request and get the root object, and *then* ask for the path we want, turning our beautiful single-round protocol into a two RTT protocol

Reply:
Yup. You're 100% right. I think the go-selector functions that were proposed included a node and I went off that, but it will have to be CID.

Will change.

Reply:
changed, fyi.